contracts/pallet-ibc/src/light_clients.rs

```diff
@@ -75,10 +84,76 @@ impl tendermint_light_client_verifier::host_functions::HostFunctionsProvider
 
 impl ics07_tendermint::HostFunctionsProvider for HostFunctionsManager {}
 
+pub struct GrandpaHederHashesStorageInstance;
+impl StorageInstance for GrandpaHederHashesStorageInstance {
+	fn pallet_prefix() -> &'static str {
+		"ibc.lightclients.grandpa"
+	}
+
+	const STORAGE_PREFIX: &'static str = "HeaderHashes";
+}
+pub type GrandpaHeaderHashesStorage = StorageValue<
+	GrandpaHederHashesStorageInstance,
+	BoundedVec<H256, ConstU32<GRANDPA_BLOCK_HASHES_CACHE_SIZE>>,
+	ValueQuery,
+>;
+
+pub struct GrandpaHeaderHashesSetStorageInstance;
+impl StorageInstance for GrandpaHeaderHashesSetStorageInstance {
+	fn pallet_prefix() -> &'static str {
+		"ibc.lightclients.grandpa"
+	}
+
+	const STORAGE_PREFIX: &'static str = "HeaderHashesSet";
+}
+pub type GrandpaHeaderHashesSetStorage = StorageValue<
+	GrandpaHeaderHashesSetStorageInstance,
+	BoundedBTreeSet<H256, ConstU32<GRANDPA_BLOCK_HASHES_CACHE_SIZE>>,
+	ValueQuery,
+>;
+
+/// Maximum number of block number to block hash mappings to keep (oldest pruned first).
+const GRANDPA_BLOCK_HASHES_CACHE_SIZE: u32 = 500;
+
 impl grandpa_client_primitives::HostFunctions for HostFunctionsManager {
+	type Header = RelayChainHeader;
+
 	fn ed25519_verify(sig: &ed25519::Signature, msg: &[u8], pub_key: &ed25519::Public) -> bool {
 		pub_key.verify(&msg, sig)
 	}
+
+	fn insert_relay_header_hashes(new_hashes: &[<Self::Header as Header>::Hash]) {
+		if new_hashes.is_empty() {
+			return
+		}
+
+		GrandpaHeaderHashesSetStorage::mutate(|hashes_set| {
```

Comment:
why can't we use only one of  `GrandpaHeaderHashesSetStorage` or `GrandpaHeaderHashesStorage`, they are storing the same values

Reply:
Because we need to keep order of the headers (to remove the oldest ones)

Reply:
But BoundedVec would always keep the items in the order of insertion.

Reply:
Yes, but we can't access an element in `BoundedVec` in O(1) by the hash

Reply:
Yeah that's true.